wp-content/themes/borderzine/functions.php

```diff
@@ -101,4 +101,80 @@ function borderzine_has_avatar( $email ) {
 		}
 	}
 	return false;
+}
+
+/**
+ * Modifies the default WordPress search query to see if the 
+ * search terms match any author names and return any posts found by said author
+ * 
+ * @param String $posts_search SQL query for the search
+ * @param Object $wp_query_obj The current wp_query object
+ * @return String $posts_search SQL query for the search
+ * @since 0.1
+ */
+function borderzine_search_posts_by_author( $posts_search, $wp_query_obj ) {
+
+	if ( !is_search() || empty( $posts_search ) ) {
+		return $posts_search;
+	}
+
+	global $wpdb;
+
+	// search all authors to see if any having a name that matches the search term
+	$search = sanitize_text_field( get_query_var( 's' ) );
+	$args = array(
+		'count_total' => false,
+		'search' => sprintf( '*%s*', $search ),
+		'fields' => 'ID',
+		'has_published_posts' => true,
+	);
+	$matching_users = get_users( $args );
+
+	// don't modify the query if there aren't any matching users
+	if ( empty( $matching_users ) ) {
+		return $posts_search;
+	}
+
+	// count matching users
+	$matching_users_count = count( $matching_users );
+
+	// prepare array with correct amount of placeholders that match the user count
+	$matching_users_placeholders = array_fill( 0, $matching_users_count, '%s' );
+
+	// put all matching user ids into one string
+	$matching_users_str = implode( ', ', $matching_users_placeholders );
+
+	// use wpdb to prepare our partial query
+	$posts_search_author_query = 
+		$wpdb->prepare(
+			"OR ( ".$wpdb->posts.".post_author IN ( $matching_users_str )))",
+			$matching_users
+		);
+
+	// modify our posts search query to also search for posts with matching user
+	$posts_search = str_replace( ')))', ")) ".$posts_search_author_query, $posts_search );
+
+	return $posts_search;
+
+}
+add_filter( 'posts_search', 'borderzine_search_posts_by_author', 10, 2 );
+
+/**
+ * Finds author ids by their display name
+ * 
+ * @param String $display_name The display name to search for
+ * @return String $user->ID The ID of the matched author
+ */
+function get_author_id_by_display_name( $display_name ) {
+
+	global $wpdb;
+	
+	$display_name = sanitize_text_field( $display_name );
+
+    if( !$user = $wpdb->get_row( $wpdb->prepare( "SELECT `ID` FROM $wpdb->users WHERE `display_name` = %s", $display_name ) ) ){
```

Comment:
Should we try to cache this? https://codex.wordpress.org/Class_Reference/WP_Object_Cache

Reply:
_Maybe_, but I'd say let's hold off on that discussion until we get an answer about whether or not we want it to return multiple authors.

Reply:
@benlk so now that this includes searching for multiple authors, here are the questions I have about caching the query:

- [ ] When should the object cache expire?
- [ ] When should it be overwritten?
- [ ] What prefix should we use for the key? `borderzine_search_display_name_*`?

Reply:
Caching this would only cache the list of IDs matching user names, so would it make sense to keep the cache for a week/month? We'd need to expire all the user-name-search caches when a user has been updated, but that cache would need to hit *all* user-name-search caches.

I initially said the Object Cache, but since we'd want to persist this across page loads, this would be better saved in the [Transients API](https://codex.wordpress.org/Transients_API).

Having separate prefixed keys makes sense from the point of view of reducing overhead when we request things from the DB, but makes it harder to delete all user-name-search caches. We can get around that by:

1. Putting all name searches in one transient (potential for absolutely huge db entry, as we're running this user search on every search), and deleting that transient on expiry, **or**
2. Putting each successful name search in its own transient (not caching unsuccessful ones), and adding the transient keys there to a transient that exists to hold that list of transient keys. When a user is created, updated, or deleted, we'd then look up the transient-of-transients, delete every one of the saved-successful searches, and then delete the transient-of-transients. (Maybe deletion of the transient-of-transients could be queued for WP_Cron or a background task, so that it doesn't affect the frontend when saving the user.)

The second option is more complicated, but seems like it would cause lower hits on page load, especially if we don't set the individual search transients or the transient-of-transients to autoload.

Reply:
What are the pros and cons for this? @joshdarby

Reply:
And, can we deploy what we have without the caching and explore that part later? or does it need to be done all at once?